File: go/heap/215_findKthLargest.go
package heap

import "container/heap"

// findKthLargest 找到数组中第k大的元素, Miniheap
func findKthLargest(nums []int, k int) int {
	// heapInt define in 40_offer_get_leatest_nums

	h := make(heapInt, 0)

	// 构建长度为k 的小根堆
	heap.Init(&h)

	for _, v := range nums {
		heap.Push(&h, v)
		if h.Len() > k {
			heap.Pop(&h)
		}
	}

	return h[0]
}

File: rust/src/heap/_215_find_kth_largest.rs
/// 构建一个长度为k 的小根堆，比k大的元素进入小根堆中，k就是最小的元素
pub fn find_kth_largest(nums: Vec<i32>, k: i32) -> i32 {
    // 长度为k的小根堆
    use std::cmp::Reverse;
    use std::collections::BinaryHeap;

    let mut heap = BinaryHeap::with_capacity(k as usize);

    for item in nums {
        if heap.len() == heap.capacity() {
            // Reverse 反转比较器，Reverse(i) > Reverse(j) means i < j
            // 找到 第k 大的元素，长度为k 的小根堆，堆顶就是需要的k
            // 所以容量满了以后，比新元素item 比堆顶大了之后 可以入堆，即Reverse(item) < old
            if let Some(old) = heap.peek() {
                if Reverse(item) < *old {
                    // pop 然后入堆
                    heap.pop();
                } else {
                    // 不需要入堆
                    continue;
                }
            }
        }

        heap.push(Reverse(item));
    }

    heap.peek().unwrap().0
}
